sacred/observers/s3_observer.py

```diff
@@ -0,0 +1,328 @@
+                               file_location).put(Body=open(filename, 'rb'))
+
+    def save_cout(self):
+        binary_data = self.cout[self.cout_write_cursor:].encode("utf-8")
+        key = os.path.join(self.dir, 'cout.txt')
+        self.put_data(key, binary_data)
+        self.cout_write_cursor = len(self.cout)
+
+    def heartbeat_event(self, info, captured_out, beat_time, result):
+        self.info = info
+        self.run_entry['heartbeat'] = beat_time.isoformat()
+        self.run_entry['result'] = result
+        self.cout = captured_out
+        self.save_cout()
+        self.save_json(self.run_entry, 'run.json')
+        if self.info:
+            self.save_json(self.info, 'info.json')
+
+    def completed_event(self, stop_time, result):
+        self.run_entry['stop_time'] = stop_time.isoformat()
+        self.run_entry['result'] = result
+        self.run_entry['status'] = 'COMPLETED'
+
+        self.save_json(self.run_entry, 'run.json')
+
+    def interrupted_event(self, interrupt_time, status):
+        self.run_entry['stop_time'] = interrupt_time.isoformat()
+        self.run_entry['status'] = status
+        self.save_json(self.run_entry, 'run.json')
+
+    def failed_event(self, fail_time, fail_trace):
+        self.run_entry['stop_time'] = fail_time.isoformat()
+        self.run_entry['status'] = 'FAILED'
+        self.run_entry['fail_trace'] = fail_trace
+        self.save_json(self.run_entry, 'run.json')
+
+    def resource_event(self, filename):
+        store_path, md5sum = self.find_or_save(filename, self.resource_dir)
+        self.run_entry['resources'].append([filename, store_path])
+        self.save_json(self.run_entry, 'run.json')
+
+    def artifact_event(self, name, filename, metadata=None, content_type=None):
+        self.save_file(filename, name)
+        self.run_entry['artifacts'].append(name)
+        self.save_json(self.run_entry, 'run.json')
+
+    def log_metrics(self, metrics_by_name, info):
+        """Store new measurements into metrics.json.
+        """
+
+        for metric_name, metric_ptr in metrics_by_name.items():
+
+            if metric_name not in self.saved_metrics:
+                self.saved_metrics[metric_name] = {"values": [],
+                                                   "steps": [],
+                                                   "timestamps": []}
+
+            self.saved_metrics[metric_name]["values"] += metric_ptr["values"]
+            self.saved_metrics[metric_name]["steps"] += metric_ptr["steps"]
+
+            timestamps_norm = [ts.isoformat()
+                               for ts in metric_ptr["timestamps"]]
+            self.saved_metrics[metric_name]["timestamps"] += timestamps_norm
+
+        self.save_json(self.saved_metrics, 'metrics.json')
+
+    def __eq__(self, other):
+        if isinstance(other, S3FileObserver):
+            return (self.bucket == other.bucket and
+                    self.basedir == other.basedir)
```

Comment:
I would put the `and` on the next line just for convention.

Reply:
If I do that, flake8 gives me an error:  W503 line break before binary operator